### PR TITLE
Support top n sigma sampling

### PR DESCRIPTION
**Pull request by Snowdar**
## Motivation
Support a more complete samling options for decoding, especialy for openai compatible api.

## Modifications

1. Add the rencent [top-n-sigma samping](https://arxiv.org/html/2411.07641v1) as a decoding option and it will be applied to the logits after temperature applied when `top_n_sigma` > 0.  The batch processing has been implemented.
2. Add `top-k`, `min-p` and `top_n_sigma` params for openai compatible api server (Note, please add these three params by `extra_body` when using openai SDK)

client example:
```
client = OpenAI(api_key='no key', base_url='http://localhost:30000/v1')
response = client.chat.completions.create(
      model='llama',
      messages=dict(role='user', content='hi'),
      temperature=0.3,
      max_tokens=128,
      top_p=0.95,
      frequency_penalty=0.,
      presence_penalty=0.,
      extra_body=dict(
          min_p=0.,
          top_k=-1,
          top_n_sigma=1, # enable it
      ),
)
```

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/openai_api/protocol.py [hunk 1]
```diff
@@ -273,6 +273,8 @@ class ChatCompletionRequest(BaseModel):
     stream_options: Optional[StreamOptions] = None
     temperature: float = 0.7
     top_p: float = 1.0
+    top_k: int = -1
```
Comment (merrymercy) [inline python/sglang/srt/openai_api/protocol.py]: move this after L288

File: python/sglang/srt/layers/sampler.py [hunk 2]
```diff
@@ -50,6 +50,20 @@ def forward(
         else:
             # Post process logits
             logits.div_(sampling_info.temperatures)
+
+            if any(sampling_info.top_n_sigmas > 0):
```
Comment (merrymercy) [inline python/sglang/srt/layers/sampler.py]: move this check to `sampling_batch_info.py` and name it `need_top_n_sigmas_sampling`, similar to how we do `need_min_p_sampling`.

File: python/sglang/srt/layers/sampler.py [hunk 3]
```diff
@@ -50,6 +50,20 @@ def forward(
...
+                sigma = torch.std(logits, dim=-1, keepdim=True)
+                # Create mask and enable only for the requests that have top_n_sigma > 0
+                mask = (sampling_info.top_n_sigmas.view(-1, 1) <= 0) | (
+                    logits >= max_logit - sampling_info.top_n_sigmas.view(-1, 1) * sigma
+                )
+
+                # Apply mask
+                logits = torch.where(
+                    mask, logits, torch.tensor(float("-inf")).to(logits.device)
+                )
```
Comment (merrymercy) [inline python/sglang/srt/layers/sampler.py]: To prevent this function from being too long, wrap this as a function and call it here.

**Review by merrymercy — CHANGES_REQUESTED** [review]
Thanks for the contribution. I left a few comments.

**Comment by Snowdar** [comment]
> Thanks for the contribution. I left a few comments.

OK. All of these have completed. Could you please view it again?

**Comment by merrymercy** [comment]
`top-k` and `min-p` are fixed in https://github.com/sgl-project/sglang/pull/2499.
However, top n sigma is still not a common sampling strategy, we may not want to maintain it in the core right now.
For these things, it is better to support them via a custom logit processer interface. Can you help review this issue? https://github.com/sgl-project/sglang/issues/2291

I will close this for now.

**Comment by Snowdar** [comment]
> `top-k` and `min-p` are fixed in #2499. However, top n sigma is still not a common sampling strategy, we may not want to maintain it in the core right now. For these things, it is better to support them via a custom logit processer interface. Can you help review this issue? #2291
> 
> I will close this for now.

So if the custom logit processer could be supported，then any other sampling codes will not be merged into sglang (considering sharing the custom samping to other people by the sglang itself)?

**Comment by merrymercy** [comment]
Hi @Snowdar ,

Custom logit processor is supported by https://github.com/sgl-project/sglang/pull/2396. We can implement top n sigma with the custom logit processor.

We can still accept new sampling methods. We can merge some built-in custom logit processor into the sglang source code. Contributions are welcome!

**Comment by Snowdar** [comment]
> Hi @Snowdar ,
> 
> Custom logit processor is supported by #2396. We can implement top n sigma with the custom logit processor.
> 
> We can still accept new sampling methods. We can merge some built-in custom logit processor into the sglang source code. Contributions are welcome!

OK. Good job! I will try this new method.

**Comment by hongpeng-guo** [comment]
> OK. Good job! I will try this new method.

Here is an example of the custom logit processor :) https://docs.sglang.ai/references/sampling_params.html#custom-logit-processor
